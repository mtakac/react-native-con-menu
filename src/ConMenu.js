import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import {
    StyleSheet,
    Text,
    View,
    Button,
    TouchableNativeFeedback,
    Modal,
    TouchableOpacity
} from 'react-native';

import AnimatedView from './AnimatedView';

const DEFAULT_ANIMATION_DURATION = 250;
const DEFAULT_MENU_WIDTH = 200;
const DEFAULT_ITEM_COLOR = '#EEEEEE';
const DEFAULT_ITEM_PADDING = 20;
const DEFAULT_ITEM_PRESSED_COLOR = '#DDDDDD';
const DEFAULT_TOGGLE_TEXT = 'Menu';

const propTypes = {
    items: PropTypes.arrayOf(
        PropTypes.shape({
            label: PropTypes.string.isRequired,
            onPress: PropTypes.func
        })
    ).isRequired,

    animationDuration: PropTypes.number,
    itemPressedColor: PropTypes.string,
    position: PropTypes.shape({ top: PropTypes.number, left: PropTypes.number }),
    menuStyle: PropTypes.shape({ width: PropTypes.number.isRequired }),
    itemStyle: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
    toggleText: PropTypes.string,
    ToggleComponent: PropTypes.func
};

const defaultProps = {
    animationDuration: DEFAULT_ANIMATION_DURATION,
    itemPressedColor: DEFAULT_ITEM_PRESSED_COLOR,
    position: { top: 0, right: 0 },
    menuStyle: { width: DEFAULT_MENU_WIDTH },
    itemStyle: {},
    toggleText: DEFAULT_TOGGLE_TEXT,
    ToggleComponent: null
};

class ConMenu extends Component {
    timeOut = null;
    buttonView = createRef();

    state = {
        isMounted: false,
        isModalOpen: false,
        top: 0,
        right: 0
    };

    componentDidUpdate(prevProps, prevState) {
        const { animationDuration } = this.props;
        const { isMounted } = this.state;

        if (prevState.isMounted && !isMounted) {
            this.timeOut = setTimeout(
                () => this.setState({ isModalOpen: false }),
                animationDuration
            );
        }

        if (!prevState.isMounted && isMounted) {
            this.setState({ isModalOpen: true });
        }
    }

    componentWillUnmount() {
        clearTimeout(this.timeOut);
    }

    toggle = () => {
        this.setState(state => ({ ...state, isMounted: !state.isMounted }));
    };

    measureButton = () => {
        const { menuStyle } = this.props;

        this.buttonView.current.measure((x, y, width, height, pageX, pageY) => {
            this.setState({ top: pageY + height, left: pageX - menuStyle.width + width });
        });
    };

    renderToggleComponent = () => {
        const { ToggleComponent, toggleText } = this.props;

        if (ToggleComponent) {
            return <ToggleComponent onPress={this.toggle} />;
        }

        return <Button title={toggleText} onPress={this.toggle} />;
    };

    render() {
        const {
            items,
            animationDuration,
            itemPressedColor,
            position,
            menuStyle,
            itemStyle
        } = this.props;

        const { isModalOpen, isMounted, top, left } = this.state;

        return (
            <View style={{ ...styles.container, ...position }}>
                <View onLayout={this.measureButton} ref={this.buttonView}>
                    {this.renderToggleComponent()}
                </View>

                <Modal visible={isModalOpen} animationType="none" transparent>
                    <TouchableOpacity activeOpacity={1} onPressOut={this.toggle}>
                        <AnimatedView
                            style={styles.animatedView}
                            duration={animationDuration}
                            isMounted={isMounted}
                        >
                            <View style={{ ...menuStyle, ...styles.menu, top, left }}>
                                {items.map(({ label, onPress }) => (
                                    <TouchableNativeFeedback
                                        key={label}
                                        background={TouchableNativeFeedback.Ripple(
                                            itemPressedColor
                                        )}
                                        onPress={onPress}
                                    >
                                        <View style={{ ...styles.item, ...itemStyle }}>
                                            <Text>{label}</Text>
                                        </View>
                                    </TouchableNativeFeedback>
                                ))}
                            </View>
                        </AnimatedView>
                    </TouchableOpacity>
                </Modal>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        position: 'absolute'
    },

    animatedView: {
        width: '100%',
        height: '100%'
    },

    menu: {
        position: 'absolute'
    },

    item: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        padding: DEFAULT_ITEM_PADDING,
        backgroundColor: DEFAULT_ITEM_COLOR
    }
});

ConMenu.propTypes = propTypes;
ConMenu.defaultProps = defaultProps;

export default ConMenu;
