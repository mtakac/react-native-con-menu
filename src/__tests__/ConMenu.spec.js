import React from 'react';
import { TouchableNativeFeedback, Text } from 'react-native';
import { shallow } from 'enzyme';
import ConMenu from '../ConMenu';

describe('<ConMenu />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<ConMenu items={[]} />);
    });

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('renders items correctly', () => {
        const items = [
            {
                label: 'Menu item',
                onPress: () => {}
            }
        ];
        wrapper.setProps({ items });
        expect(wrapper).toMatchSnapshot();
    });

    it('renders <ToggleComponent /> correctly', () => {
        const ToggleComponent = props => (
            <TouchableNativeFeedback onPress={props.onPress}>
                <Text>Press me!</Text>
            </TouchableNativeFeedback>
        );

        wrapper.setProps({ ToggleComponent });
        expect(wrapper).toMatchSnapshot();
    });
});
