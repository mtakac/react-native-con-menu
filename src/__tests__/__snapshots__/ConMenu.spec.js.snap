// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConMenu /> renders <ToggleComponent /> correctly 1`] = `
<Component
  style={
    Object {
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <Component
    onLayout={[Function]}
  >
    <ToggleComponent
      onPress={[Function]}
    />
  </Component>
  <Component
    animationType="none"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <TouchableOpacity
      activeOpacity={1}
      onPressOut={[Function]}
    >
      <_class2
        duration={250}
        isMounted={false}
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      >
        <Component
          style={
            Object {
              "left": undefined,
              "position": "absolute",
              "top": 0,
              "width": 200,
            }
          }
        />
      </_class2>
    </TouchableOpacity>
  </Component>
</Component>
`;

exports[`<ConMenu /> renders correctly 1`] = `
<Component
  style={
    Object {
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <Component
    onLayout={[Function]}
  >
    <Button
      onPress={[Function]}
      title="Menu"
    />
  </Component>
  <Component
    animationType="none"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <TouchableOpacity
      activeOpacity={1}
      onPressOut={[Function]}
    >
      <_class2
        duration={250}
        isMounted={false}
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      >
        <Component
          style={
            Object {
              "left": undefined,
              "position": "absolute",
              "top": 0,
              "width": 200,
            }
          }
        />
      </_class2>
    </TouchableOpacity>
  </Component>
</Component>
`;

exports[`<ConMenu /> renders items correctly 1`] = `
<Component
  style={
    Object {
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <Component
    onLayout={[Function]}
  >
    <Button
      onPress={[Function]}
      title="Menu"
    />
  </Component>
  <Component
    animationType="none"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <TouchableOpacity
      activeOpacity={1}
      onPressOut={[Function]}
    >
      <_class2
        duration={250}
        isMounted={false}
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      >
        <Component
          style={
            Object {
              "left": undefined,
              "position": "absolute",
              "top": 0,
              "width": 200,
            }
          }
        >
          <DummyTouchableNativeFeedback
            background={Object {}}
            key="Menu item"
            onPress={[Function]}
          >
            <Component
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#EEEEEE",
                  "flex": 1,
                  "justifyContent": "center",
                  "padding": 20,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              >
                Menu item
              </Text>
            </Component>
          </DummyTouchableNativeFeedback>
        </Component>
      </_class2>
    </TouchableOpacity>
  </Component>
</Component>
`;
